### PR TITLE
Remove example, make Nginx config extensible, simplify configurations

### PR DESCRIPTION
**Pull request by jasonpincin**
The overall goal of this PR is to both simplify the blueprint, and make it more extensible. Here's the high level changes:
- Removed all traces of example. Examples to be done outside this repo, eventually replaced with tests.
- Split Nginx config into multiple files: `nginx.conf`, `ssl.conf`, `health.conf`, and `conf.d/site.conf`
- Multiple script changes to increase clarity of containerpilot.json
  - One ACME task (`acme checkin`) instead of three
  - `acme watch` handles details of launching consul-template daemon correctly 
  - Broke `reload.sh` out into `generate-config` and `reload`

With these changes, `containerpilot.json` is more comprehensible, and more importantly, users can extend the Nginx configuration in pieces by adding/replacing files in `/etc/nginx/conf.d`. Most users will simply need to replace `/etc/nginx/conf.d/site.conf` with a bare-minimum set of directives tailored for the application to get things up and running, without having to worry about SSL details or health check details, etc. 

Todo:
- Update README - describe available env vars `CONSUL`, `CONSUL_AGENT`, `ACME_DOMAIN`, `ACME_ENV`, as well as best practices on extending Nginx config and `containerpilot.json`

### CODE REVIEW

**Comment by tgross** [comment]
> Removed all traces of example. Examples to be done outside this repo, eventually replaced with tests.

One of our stated objectives all along has been for each application blueprint to be standalone so that they can act as demonstrations of the principles we're trying to illustrate. Removing the example means that you have to use this with some other blueprint, which increases the barrier to entry for a demonstration. (Plus it'll frustrate Joyent sales folks who want to have quick-and-ready demos that don't take 10-15 minutes to stand up.)

File: etc/containerpilot.json [hunk 1]
```diff
@@ -49,9 +49,7 @@
     {{ if and .CONSUL_AGENT .ACME_DOMAIN }},{{ end }}
     {{ if .ACME_DOMAIN }}
     {
-      "command": ["/usr/local/bin/consul-template",
-                  "-config", "/etc/acme/watch.hcl",
-                  "-consul", "{{ if .CONSUL_AGENT }}localhost{{ else }}{{ if .CONSUL }}{{ .CONSUL }}{{ else }}consul{{ end }}{{ end }}:8500"],
+      "command": ["acme", "watch"],
```
Comment (tgross) [inline etc/containerpilot.json]: Why not just use `/usr/local/bin/consul-template -config /etc/acme/watch.hcl -consul $CONSUL_HOST:8500` directly here?

File: etc/containerpilot.json [hunk 2]
```diff
@@ -1,43 +1,43 @@
...
-      "health": "/usr/bin/curl --fail --silent --show-error --output /dev/null http://localhost/nginx-health",
+      "health": "curl -fsSo /dev/null http://localhost/nginx-health",
       "poll": 10,
       "ttl": 25,
       "interfaces": ["eth1", "eth0"]
     }{{ if .ACME_DOMAIN }},
     {
+      "name": "nginx-ssl",
+      "port": 443,
+      "health": "acme init && curl -kfsSo /dev/null https://localhost/nginx-health",
```
Comment (tgross) [inline etc/containerpilot.json]: If we put `-k` here then we're intentionally bypassing the SSL check. I feel like it'd be really useful for the health check to include verifying SSL. Is it feasible to switch out `localhost` for `{{ .ACME_DOMAIN }}` here so that we can ensure SSL is working in the health check?

File: etc/nginx/ssl.conf [hunk 3]
```diff
@@ -0,0 +1,10 @@
+ssl_certificate /var/www/ssl/fullchain.pem;
+ssl_certificate_key /var/www/ssl/privkey.pem;
+ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
+ssl_prefer_server_ciphers on;
+ssl_ciphers 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA';
```
Comment (tgross) [inline etc/nginx/ssl.conf]: I'd like to recommend that we put a link to where this configuration came from in a comment, so that a) we can occasionally check on it to see if it needs updating, and b) we lend a bit of community-approval legitimacy to the SSL config. Maybe another comment would be that it generates a "A" rating, etc.

File: bin/generate-config [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+CERT_DIR="/var/www/ssl"
+
+# Determine whether ACME SSL certificates have been acquired
+SSL_READY="false"
+if [ -f ${CERT_DIR}/fullchain.pem -a -f ${CERT_DIR}/privkey.pem ]; then
+    export SSL_READY="true"
+fi
+
+# Generate a conf.d config file for every corresponding cond.d Consul template
+for f in $(ls -1 /etc/nginx/templates/conf.d/)
```
Comment (tgross) [inline bin/generate-config]: Not sure if you missed it in the [consul-template docs](https://github.com/hashicorp/consul-template) but we can render multiple files from a single invocation, which should reduce the number of API calls to Consul. Their example is as follows:

```
$ consul-template \
  -consul my.consul.internal:6124 \
  -template "/tmp/nginx.ctmpl:/var/nginx/nginx.conf:service nginx restart" \
  -template "/tmp/redis.ctmpl:/var/redis/redis.conf:service redis restart" \
  -template "/tmp/haproxy.ctmpl:/var/haproxy/haproxy.conf"
```

File: etc/containerpilot.json (hunk 1 above)
Comment (jasonpincin) [inline etc/containerpilot.json]: We certainly can. I just did it this way so that all the moving parts related to certificate acquisition was contained within the `acme` script, so that when you're looking at that you see it all, versus some pieces here, some there. Things like the `CONSUL_AGENT` are tucked away cleanly in the `acme` script, etc. Cleans up `containerpilot.json` a bit. Admittedly, this is purely stylistic.

File: etc/containerpilot.json (hunk 2 above)
Comment (jasonpincin) [inline etc/containerpilot.json]: Yeah, I agree. The real issue here is `ACME_ENV`. When set to `staging`, `-k` is required because we're dealing with test, invalid certs. When we're dealing with `production` as a value of that environment, then we could swap the host, although that feels a little risky as there's no guarantee that DNS is resolving to our container, etc... not sure off the top of my head whether just passing `Host:` header would be enough. Handling all these cases may be complex enough to warrant moving the test into a script instead of cramming all the branches into `containerpilot.json`, thoughts?

File: etc/containerpilot.json (hunk 1 above)
Comment (tgross) [inline etc/containerpilot.json]: Ok, that's totally reasonable. Let's keep it as you've got it here.

File: etc/containerpilot.json (hunk 2 above)
Comment (tgross) [inline etc/containerpilot.json]: Ah, it's the staging bit that I was forgetting here. Does LE have a recommended way to test the staging cert? Like maybe if there's a signing root that we can include in our container we can test locally safely?

> Handling all these cases may be complex enough to warrant moving the test into a script instead of cramming all the branches into containerpilot.json, thoughts?

That's probably not a bad idea if we need to implement a lot of logic. We've found that tends to be the case for a lot of the blueprints as we start probing all their edge cases!

Comment (jasonpincin) [inline etc/containerpilot.json]: > Does LE have a recommended way to test the staging cert?

Good question; I'll look into that.

Comment (jasonpincin) [inline etc/containerpilot.json]: >  Is it feasible to switch out localhost for {{ .ACME_DOMAIN }} here so that we can ensure SSL is working in the health check?

Bit of a chicken and egg here, as the health check must pass prior to the service registering and accepting requests for the domain, even if we assume DNS is correct. Perhaps we append a line to /etc/hosts:

127.0.0.1 $ACME_DOMAIN

... prior to test...

**Comment by tgross** [comment]
Carrying all these commits in https://github.com/autopilotpattern/nginx/pull/41. Closing this PR.
